### PR TITLE
KYLIN-4315 use metadata numRows in beeline client for quick row counting

### PR DESCRIPTION
**Pull request by xiacongling**
see https://issues.apache.org/jira/browse/KYLIN-4315

### CODE REVIEW

File: source-hive/src/main/java/org/apache/kylin/source/hive/BeelineHiveClient.java [hunk 1]
```diff
@@ -272,6 +261,9 @@ private void extractTableParam(ResultSet resultSet, HiveTableMetaBuilder builder
             if ("numFiles".equals(resultSet.getString(2).trim())) {
                 builder.setFileNum(Long.parseLong(resultSet.getString(3).trim()));
             }
+            if ("numRows".equals(resultSet.getString(2).trim())) {
+                builder.setRowNum(Long.parseLong(resultSet.getString(3).trim()));
```
Comment (nichunen) [inline source-hive/src/main/java/org/apache/kylin/source/hive/BeelineHiveClient.java]: Is it certain that "numRows" will always be gotten in hive table's meta? As I know, there is possibility that it be -1

**Comment by RupengWang** [comment]
### Check list
- [x] Step 1. [**OPTIONAL**]Understand background/root cause/design basically(one hour EST) . Its issue type is 
    - [ ] bug fix
    - [ ] new feature
    - [x] enhancement
- [x] Step 2. Test cases are designed and documented (30 minutes EST).
- [x] Step 3. Prepare specific env, for example:
    - mock data (maybe some specific data type)
    - test env (maybe install a RDBMS instance).
- [x] Step 4. Verify and make sure test cases passed.
- [x] Step 5. Paste manual important steps and screenshots here (20 minutes EST). 
    - If you find difficulty in pick most important evidence, please attach diagnosis package. 
- [x] Step 6. Do more check in user perspective (20 minutes EST)
    - [x] Doc need be updated? And if it is updated? Ask help for release manager if so.
    - [x] If it is a breaking change so we should notify Kylin community? Ask help for release manager if so.
- [x] Step 7.  Summarize this test (20 minutes EST).

### Total estimate
- 4 hours for small issue from optimistic estimate. 

### Note
If you find some unexpected and **unrelated** error/mistake, please DO report it if it is truly a mistake, I think we may research and fix it in the future.

If you find background information/root cause analysis is not complete or ambiguous, please try to contact to author or do a quick research and record what you find. It is a good chance to learn something interesting.

If you find it is not easy to design testcase, please notify release manager.

**Comment by RupengWang** [comment]
## Design test cases

### Prepare test env
- create cube and build segment (2012-01-01 ~ 2013-01-01)
  - dimension : LSTG_FORMAT_NAME, ops_region, part_dt
  - measure : sum(price)

### Case-1

Make sure query-1 and query-2 has the same result.

- query-1 (Hive)
```sql
select LSTG_FORMAT_NAME, stddev_pop(SUM_M)
from (
      select LSTG_FORMAT_NAME, ops_region, sum(price) as SUM_M
      from KYLIN_SALES
      where part_dt >= '2012-01-01' and part_dt < '2012-06-01'
      group by LSTG_FORMAT_NAME, ops_region
) a
group by LSTG_FORMAT_NAME
order by LSTG_FORMAT_NAME
```

- query-2 (Kylin)
```sql
select LSTG_FORMAT_NAME, stddev_sum(ops_region, price)
from KYLIN_SALES
where part_dt >= '2012-01-01' and part_dt < '2012-06-01'
group by LSTG_FORMAT_NAME
order by LSTG_FORMAT_NAME
```

### Case-2

Make sure query-1 and query-2 has the same result.

- query-1 (Hive)
```sql
select stddev_pop(SUM_M)
from (
      select LSTG_FORMAT_NAME, ops_region, sum(price) as SUM_M
      from KYLIN_SALES
      where part_dt >= '2012-01-01' and part_dt < '2012-06-01'
      group by LSTG_FORMAT_NAME, ops_region
) a
```

- query-2(Kylin)
```sql
select stddev_sum(ops_region, price)
from KYLIN_SALES
where part_dt >= '2012-01-01' and part_dt < '2012-06-01'
```

**Comment by RupengWang** [comment]
## Prepare data and env

- [x] Built on commit_SHA : d14eef5fee3ef5a305b35b9121bde28fc2400d9d
- [x] Hadoop env : CDH 5.7

- [x] Just sample cube is OK.
image

**Comment by RupengWang** [comment]
### Verify test case
#### Case 1
1. Result of Kylin
image
2. Result of Hive
```sql
DROP TABLE IF EXISTS kylin_intermediate_kylin_sales_cube_f91987ee_443f_cadb_f3cb_04accea358b6;
CREATE EXTERNAL TABLE IF NOT EXISTS kylin_intermediate_kylin_sales_cube_f91987ee_443f_cadb_f3cb_04accea358b6
(
`KYLIN_SALES_TRANS_ID` bigint
,`KYLIN_SALES_PART_DT` date
,`KYLIN_SALES_LEAF_CATEG_ID` bigint
,`KYLIN_SALES_LSTG_SITE_ID` int
,`KYLIN_CATEGORY_GROUPINGS_META_CATEG_NAME` string
,`KYLIN_CATEGORY_GROUPINGS_CATEG_LVL2_NAME` string
,`KYLIN_CATEGORY_GROUPINGS_CATEG_LVL3_NAME` string
,`KYLIN_SALES_LSTG_FORMAT_NAME` string
,`KYLIN_SALES_SELLER_ID` bigint
,`KYLIN_SALES_BUYER_ID` bigint
,`BUYER_ACCOUNT_ACCOUNT_BUYER_LEVEL` int
,`SELLER_ACCOUNT_ACCOUNT_SELLER_LEVEL` int
,`BUYER_ACCOUNT_ACCOUNT_COUNTRY` string
,`SELLER_ACCOUNT_ACCOUNT_COUNTRY` string
,`BUYER_COUNTRY_NAME` string
,`SELLER_COUNTRY_NAME` string
,`KYLIN_SALES_OPS_USER_ID` string
,`KYLIN_SALES_OPS_REGION` string
,`KYLIN_SALES_PRICE` decimal(19,4)
)
STORED AS SEQUENCEFILE
LOCATION 'hdfs://sandbox.hortonworks.com:8020/kylin/kylin_metadata_wrp_310/kylin-c9804baf-044f-4537-132a-81ce89d76e10/kylin_intermediate_kylin_sales_cube_f91987ee_443f_cadb_f3cb_04accea358b6';
ALTER TABLE kylin_intermediate_kylin_sales_cube_f91987ee_443f_cadb_f3cb_04accea358b6 SET TBLPROPERTIES('auto.purge'='true');
INSERT OVERWRITE TABLE `kylin_intermediate_kylin_sales_cube_f91987ee_443f_cadb_f3cb_04accea358b6` SELECT
`KYLIN_SALES`.`TRANS_ID` as `KYLIN_SALES_TRANS_ID`
,`KYLIN_SALES`.`PART_DT` as `KYLIN_SALES_PART_DT`
,`KYLIN_SALES`.`LEAF_CATEG_ID` as `KYLIN_SALES_LEAF_CATEG_ID`
,`KYLIN_SALES`.`LSTG_SITE_ID` as `KYLIN_SALES_LSTG_SITE_ID`
,`KYLIN_CATEGORY_GROUPINGS`.`META_CATEG_NAME` as `KYLIN_CATEGORY_GROUPINGS_META_CATEG_NAME`
,`KYLIN_CATEGORY_GROUPINGS`.`CATEG_LVL2_NAME` as `KYLIN_CATEGORY_GROUPINGS_CATEG_LVL2_NAME`
,`KYLIN_CATEGORY_GROUPINGS`.`CATEG_LVL3_NAME` as `KYLIN_CATEGORY_GROUPINGS_CATEG_LVL3_NAME`
,`KYLIN_SALES`.`LSTG_FORMAT_NAME` as `KYLIN_SALES_LSTG_FORMAT_NAME`
,`KYLIN_SALES`.`SELLER_ID` as `KYLIN_SALES_SELLER_ID`
,`KYLIN_SALES`.`BUYER_ID` as `KYLIN_SALES_BUYER_ID`
,`BUYER_ACCOUNT`.`ACCOUNT_BUYER_LEVEL` as `BUYER_ACCOUNT_ACCOUNT_BUYER_LEVEL`
,`SELLER_ACCOUNT`.`ACCOUNT_SELLER_LEVEL` as `SELLER_ACCOUNT_ACCOUNT_SELLER_LEVEL`
,`BUYER_ACCOUNT`.`ACCOUNT_COUNTRY` as `BUYER_ACCOUNT_ACCOUNT_COUNTRY`
,`SELLER_ACCOUNT`.`ACCOUNT_COUNTRY` as `SELLER_ACCOUNT_ACCOUNT_COUNTRY`
,`BUYER_COUNTRY`.`NAME` as `BUYER_COUNTRY_NAME`
,`SELLER_COUNTRY`.`NAME` as `SELLER_COUNTRY_NAME`
,`KYLIN_SALES`.`OPS_USER_ID` as `KYLIN_SALES_OPS_USER_ID`
,`KYLIN_SALES`.`OPS_REGION` as `KYLIN_SALES_OPS_REGION`
,`KYLIN_SALES`.`PRICE` as `KYLIN_SALES_PRICE`
 FROM `DEFAULT`.`KYLIN_SALES` as `KYLIN_SALES`
INNER JOIN `DEFAULT`.`KYLIN_CAL_DT` as `KYLIN_CAL_DT`
ON `KYLIN_SALES`.`PART_DT` = `KYLIN_CAL_DT`.`CAL_DT`
INNER JOIN `DEFAULT`.`KYLIN_CATEGORY_GROUPINGS` as `KYLIN_CATEGORY_GROUPINGS`
ON `KYLIN_SALES`.`LEAF_CATEG_ID` = `KYLIN_CATEGORY_GROUPINGS`.`LEAF_CATEG_ID` AND `KYLIN_SALES`.`LSTG_SITE_ID` = `KYLIN_CATEGORY_GROUPINGS`.`SITE_ID`
INNER JOIN `DEFAULT`.`KYLIN_ACCOUNT` as `BUYER_ACCOUNT`
ON `KYLIN_SALES`.`BUYER_ID` = `BUYER_ACCOUNT`.`ACCOUNT_ID`
INNER JOIN `DEFAULT`.`KYLIN_ACCOUNT` as `SELLER_ACCOUNT`
ON `KYLIN_SALES`.`SELLER_ID` = `SELLER_ACCOUNT`.`ACCOUNT_ID`
INNER JOIN `DEFAULT`.`KYLIN_COUNTRY` as `BUYER_COUNTRY`
ON `BUYER_ACCOUNT`.`ACCOUNT_COUNTRY` = `BUYER_COUNTRY`.`COUNTRY`
INNER JOIN `DEFAULT`.`KYLIN_COUNTRY` as `SELLER_COUNTRY`
ON `SELLER_ACCOUNT`.`ACCOUNT_COUNTRY` = `SELLER_COUNTRY`.`COUNTRY`
WHERE 1=1 AND (`KYLIN_SALES`.`PART_DT` >= '2012-01-01' AND `KYLIN_SALES`.`PART_DT` < '2012-02-01')
;
select count(*) from kylin_intermediate_kylin_sales_cube_f91987ee_443f_cadb_f3cb_04accea358b6;
```

**Pull request merged**